### ISSUE TITLE
very large `arithmetic-shift` crashes racketcs

### ISSUE THREAD

**Issue by samth** [racket/racket#2865]
```
[samth@huor:~/.../extra-pkgs/typed-racket/typed-racket-test (master) plt] racketcs
Welcome to Racket v7.5.0.3 [cs].
> (arithmetic-shift 2 (expt 2 200))
out of memory
```

On traditional Racket, there's also an out-of-memory error, but it returns to the prompt with an exception.

**Comment by gus-massa** [comment]
I think it's a problem in Chez Scheme. 

In the original version, when I run this

    (expand/optimize '(bitwise-arithmetic-shift 2 100000000))

it takes a lot of time. I got bored before it finished.

The problem is that `bitwise-arithmetic-shift` is marked as `mifoldable`, so the `cp0` pass try to calculate it during the compilation. I guess to fix it, it will be necessary to remove the `mifoldable` flag and add an special cases for `cp02` and more cases in `cp03`. 

All the other similar functions have the same problem.

**Comment by mflatt** [comment]
I think a handful of primitives — including `arithmetic-shift`, `make-bytes`, `make-vector`, and `make-string` — need to divert to a slow path when the size is large enough, and that slow path should check whether the size is larger than a custodian-based limit or bail out when in constant-folding mode. (That's how non-CS works.) The difficult part may be to do that without slowing down small allocations or interfering with optimization.

**Comment by mflatt** [comment]
Oh, I see that's already implemented for `make-vector`, `make-string`, and `make-bytes`. (I guess I did that and forgot.) Since `arithmetic-shift` is defined at the Rumble layer, too, maybe the repair is not so bad. I don't think we have to necessarily make Chez Scheme by itself fail more gracefully.

**Comment by gus-massa** [comment]
The difference is that in `make-vector` the allocation is easy to calculate, but in `arithmetic-shift` it is more difficult, for integers the allocation result is almost linear if the gc can clean the intermediate results.

The main problem is that the error is in the cp0 pass. For integers the **compilation** time is quadratic. Moreover this takes too long to compile:

    (expand/optimize '(if (zero? (random 1))
                         #f
                         (bitwise-arithmetic-shift 2 100000000)))

I guess it will be necessary to reimplement `bitwise-arithmetic-shift` and similar functions when it's not clear that the exponent is small.

**Comment by gus-massa** [comment]
I somewhere replaced in my head  `arithmetic-shift` with `expt`, so in the previous comment replace  every quadratic with linear. And even `expt` is linear IIUC the implementation. Anyway, it can be a long time.

**Comment by gus-massa** [comment]
I've send a PR to Chez Sheme, that should fix part of the problem. I'm not sure about the details of the fix for racketcs.

**Comment by samth** [comment]
Just to be clear, this still doesn't work in racketcs after @gus-massa's patch.

**Comment by gus-massa** [comment]
Mmm, I think I fixed the compilation time problem, but I didn't fix the runtime problem.

    #lang ChezScheme
    (expand/optimize '(ash 2 (expt 2 200)))  ;==> (#2%ash  2 160...376)

    #lang ChezScheme
    (ash 2 (expt 2 200))  ;==> <out of memory>

**Comment by mflatt** [comment]
Fixed by c48afdb16b5b3e6aac60c39ae65196eaf3d2d8b8 and racket/ChezScheme@dccc7e81b2f0909ce3c7871b849b0faa83eae576.

Also, racket/ChezScheme@8038a1745b8d176bc97c1dd69024da76626eb775 avoids freezing DrRacket with something like `(expt 3 (expt 2 100))`.
